### PR TITLE
(CDAP-1204) HBase sharded queue implementation

### PR DESCRIPTION
**Pull request by chtyim**

### CODE REVIEW

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-RBT166-19

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerState.java [hunk 1]
```diff
@@ -15,30 +15,55 @@
...
 
-  HBaseConsumerState(byte[] startRow, long groupId, int instanceId) {
+  HBaseConsumerState(ConsumerConfig consumerConfig, byte[] startRow,
+                     @Nullable byte[] barrierStartRow, @Nullable byte[] barrierEndRow) {
+    this.consumerConfig = consumerConfig;
     this.startRow = startRow;
-    this.groupId = groupId;
-    this.instanceId = instanceId;
+    this.barrierStartRow = barrierStartRow;
+    this.barrierEndRow = barrierEndRow;
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerState.java]: hmm. Naming - although I think the word barrier is appropriate - is a little strange. If I understand correctly, barrierEndRow is where the config changed, and the consumer waits for all other consumers to each this row. So I would call this simply "barrier". But what is barrierStartRow? Isn't that the barrier of the previous config change? So maybe call it previousBarrier?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java [hunk 2]
```diff
@@ -17,49 +17,122 @@
...
+  private final byte[] barrierScanStartRow;
+  private final byte[] barrierScanEndRow;
+  private Transaction transaction;
 
   HBaseConsumerStateStore(String datasetName, QueueName queueName, Table table) {
     super(datasetName, table);
     this.queueName = queueName;
     this.table = table;
-    this.minStartRow = QueueEntryRow.getQueueEntryRowKey(queueName, 0L, 0);
+    this.barrierScanStartRow = Bytes.add(queueName.toBytes(), QueueEntryRow.getQueueEntryRowKey(queueName, 0L, 0));
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java]: so this contains the queue name twice?

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java]: the row key for the barrier is actually "queueName + queueEntryRow".. which is agnostic to what the actual format of the entry row key is (the actual entry row key is "flowletId + name + writeTx + counter").

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java [hunk 3]
```diff
@@ -17,49 +17,122 @@
...
+    ConsumerState consumerState = getConsumerState(groupId, instanceId);
+    QueueBarrier startBarrier = consumerState.getStartBarrier();
+    QueueBarrier endBarrier = consumerState.getEndBarrier();
+    if (startBarrier == null && endBarrier == null) {
+      throw new IllegalStateException(
+        String.format("Unable to find barrier information for consumer. Queue: %s, GroupId: %d, InstanceId:%d",
+                      queueName, groupId, instanceId));
+    }
+
+    ConsumerGroupConfig groupConfig = startBarrier != null ? startBarrier.getGroupConfig()
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java]: I am trying to understand... Is it correct that:
- all consumers now have a state with a startBarrier (starting at 0) 
- so if the startBarrier is null, then we are in compatibility mode, consuming entries that were produced with the old strategy
- every time the config changes, a new barrier is added. That means if the state has an end barrier, then the consumer must wait for all other consumers to catch up when he reaches the barrier.

Sometimes it is useful to explain that in Javadocs or comments. The rest of the code gets much easier to understand

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java [hunk 4]
```diff
@@ -17,49 +17,122 @@
...
+                                                             : endBarrier.getGroupConfig();
+    ConsumerConfig consumerConfig = new ConsumerConfig(groupConfig, instanceId);
+    return new HBaseConsumerState(consumerConfig,
+                                  consumerState.getConsumerStartRow(),
+                                  startBarrier == null ? null : startBarrier.getStartRow(),
+                                  endBarrier == null ? null : endBarrier.getStartRow());
+  }
+
+  /**
+   * Checks if consumers in the given group has consumed all entries up to the given start row.
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java]: all consumers have consumed?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java (hunk 3 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java]: Yes it is correct. I'll add comments for that.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java [hunk 5]
```diff
@@ -240,22 +223,44 @@ public void dropAllForFlow(String namespaceId, String app, String flow) throws E
...
+      } else if (isStateStoreTable(tableId)) {
+        stateStoreTableIds.add(tableId);
+      }
+    }
+
+    // Upgrade of state store table
+    for (TableId tableId : stateStoreTableIds) {
+      LOG.info("Upgrading queue state store: {}", tableId);
+      Id.DatasetInstance stateStoreId = createStateStoreDataset(tableId.getNamespace().getId());
+      co.cask.cdap.api.dataset.DatasetAdmin admin = datasetFramework.getAdmin(stateStoreId, null);
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: why fully qualified?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueClientFactory.java [hunk 6]
```diff
@@ -120,4 +228,87 @@ private HTable createHTable(TableId tableId) throws IOException {
...
+      }
+      super.startTx(tx);
+    }
+
+    @Override
+    public void postTxCommit() {
+      super.postTxCommit();
+      HBaseQueueConsumer consumer = consumers.poll();
+      if (!consumer.isClosed()) {
+        consumers.add(consumer);
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueClientFactory.java]: So, a single consumer does not have multiple threads if it is catching up for an instance id that was removed. Instead this SmartConsumer round-robins through multiple instance ids. I am sure this simplifies the code... in terms of performance, though, catch-up to the barrier will take longer, right?

File: cdap-common/src/main/resources/cdap-default.xml [hunk 7]
```diff
@@ -424,7 +424,7 @@
 
     <property>
         <name>data.tx.snapshot.interval</name>
-        <value>300</value>
+        <value>60</value>
```
Comment (anew) [inline cdap-common/src/main/resources/cdap-default.xml]: unrelated change? why?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedFlowletInstanceUpdater.java [hunk 8]
```diff
@@ -56,16 +63,34 @@
...
+      twillController.changeInstances(flowletId, newInstanceCount).get();
+    } finally {
+      for (String id : flowlets) {
+        // Need to resume one by one due to TWILL-123.
+        // Also ignore any error as resuming a non-suspended flowlet is ok,
+        // except getting from exception log from the container.
+        try {
+          twillController.sendCommand(id, ProgramCommands.RESUME).get();
+        } catch (Exception e) {
+          LOG.warn("Failed to resume flowlet {}", id, e);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedFlowletInstanceUpdater.java]: maybe you should error out here? Because it means one flowlet is not running. So the flow is dysfunctional.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueClientFactory.java (hunk 6 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueClientFactory.java]: That's is right. It's a future improvement for using multiple-threads to catchup. The change is not that trivial as it involves changes in the Flow system.

File: cdap-common/src/main/resources/cdap-default.xml (hunk 7 above)
Comment (chtyim) [inline cdap-common/src/main/resources/cdap-default.xml]: It actually should belongs to previous PR that changes state store to use dataset. A shorter interval for taking tx snapshot is more favorable since the queue eviction coprocessor now needs to read the state store transactionally, meaning a more up-to-date tx snapshot helps.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java (hunk 2 above)
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java]: so because queueName already contains the flowlet id and name, there is some duplication. I guess that is ok.

**Comment by anew** [comment]
This is quite complex a change, but overall the code looks very good. I have only a few comments.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerState.java (hunk 1 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerState.java]: Renamed them to "previousBarrier" and "nextBarrier".

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java (hunk 4 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java]: fixed

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java (hunk 5 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: Because there is an inner class called `DatasetAdmin` as well.

**Comment by anew** [comment]
LGTM. Nice work :100:
